drivers/hmis/lib/hmis_data_cleanup/fix_enrollment_dates_20240416.rb

```diff
@@ -0,0 +1,93 @@
+  # - If exit date is in 2020, change it to 2023 but same month and day.
+  # For Bed Night Services (recordType:200) in special treatment project, where the DateProvided is in 2020
+  # - Change DateProvided to 2023 (same month and day). The project was not open in 2020
+
+  attr_accessor :special_treatment_project_id
+  def initialize(special_treatment_project_id:)
+    self.special_treatment_project_id = special_treatment_project_id
+  end
+
+  def e_t = Hmis::Hud::Enrollment.arel_table
+  def ex_t = Hmis::Hud::Exit.arel_table
+  def sv_t  = Hmis::Hud::Service.arel_table
+  def p_t = Hmis::Hud::Project.arel_table
+
+  def perform
+    puts "Fixing dates. Special treatment project: #{special_treatment_project_id}"
+    special_project = Hmis::Hud::Project.find_by(project_id: special_treatment_project_id)
+    raise 'Special project not found' unless special_project
+
+    normal_project_ids = Hmis::Hud::Project.where.not(project_id: special_treatment_project_id).pluck(:project_id)
+    fix_dates_special(special_treatment_project_id)
+    fix_dates(normal_project_ids)
+  end
+
+  def fix_dates_special(project_id)
+    puts "Updating (special): #{project_id}"
+
+    exits_to_update = Hmis::Hud::Exit.
+      joins(:enrollment, :project).
+      where(p_t[:project_id].eq(project_id)).
+      where(p_t[:project_type].in(HudUtility2024.residential_project_type_ids)).
+      where(ex_t[:exit_date].extract('year').eq(2020)).
+      where(e_t[:entry_date].gteq(ex_t[:exit_date]))
+
+    puts "Found #{exits_to_update.count} enrollments with exit date in 2020 where exit date is before entry date"
+
+    Hmis::Hud::Exit.transaction do
+      exits_to_update.each do |e|
+        new_date = e.exit_date.change(year: 2023)
+        puts "Updating exit #{e.id} exit date to #{new_date}"
+        e.exit_date = new_date
+        e.save!
+      end
+    end
+
+    services_to_update = Hmis::Hud::Service.
+      joins(:enrollment, :project).
+      where(p_t[:project_id].eq(project_id)).
+      where(p_t[:project_type].in(HudUtility2024.residential_project_type_ids)).
+      where(record_type: 200). # Bed Nights
+      where(sv_t[:date_provided].extract('year').eq(2020))
+
+    puts "Found #{services_to_update.count} bed nights with date provided in 2020"
+
+    Hmis::Hud::Service.transaction do
+      services_to_update.each do |s|
+        new_date = s.date_provided.change(year: 2023)
+        puts "Updating service #{s.id} date provided to #{new_date}"
+        s.date_provided = new_date
+        s.save!
+      end
+    end
+  end
+
+  def fix_dates(project_ids)
+    puts "Updating (normal): #{project_ids.count} projects"
+
+    exits_to_update = Hmis::Hud::Exit.
+      joins(:enrollment).
+      where(e_t[:project_id].in(project_ids)).
+      where(e_t[:entry_date].eq(ex_t[:exit_date]))
+
+    puts "Found #{exits_to_update.count} enrollments with equal entry and exit dates"
+
+    Hmis::Hud::Exit.transaction do
```

Comment:
Let me know if we should add some logging to describe whether there are any enrollments in the unexpected situations that we're _not_ correcting for in this script -- for example, exit before entry in a normal (non special treatment) project, or exit before entry in a non-residential project, or etc.

Also, while testing this out locally I was adding the line `Rails.logger.level = 1` in order to see the logs more clearly, without all the output from sql. Would that be helpful to add to the committed code?

Reply:
I think this is fine as is, we can run it with extra logging if needed